### PR TITLE
Remove the need for explicit reflection setup in bootstrap code

### LINKED ISSUE #2304: Refactor application bootstrap to hide implementation detail `ReflectionCapabilities` from the user
cc: @yjbanov , @kwalrath 

I've been talking with folks recently about the need for the user to instantiate `ReflectionCapabilities`. I remember we spoke about this a while back and that you had some concerns about hiding this from the user completely, but I can't remember the specifics and was hoping to get them documented since I get this question a lot.

The proposal is:

In application.ts
```
export function bootstrap(...) {
  if (reflector.reflectionCapabilities == null) {
    reflector.reflectionCapabilities = new ReflectionCapabilities();
  }
  bootstrapStatic(...);
}

export function bootstrapStatic(...) {
  // The current implementation of `bootstrap`
}
```

The user writes main.dart:
```
void main() {
  bootstrap(MyComponent);
}
```
which we then rewrite to:
```
import 'main.ng_deps.dart' as i0;
void main() {
  i0.initReflector();bootstrapStatic(MyComponent);
}
```

What do you think?

### PR DESCRIPTION
**Pull request by yjbanov**
Fixes #2304 

This PR introduces a small change to the bootstrap API to be used by Angular 2 applications written in Dart or those written in TypeScript and meant to transpile to Dart (such as our own example apps). Pure JS applications may continue to use the existing API.
## New API

`package:angular2/bootstrap.dart` is introduced (or for TypeScript just `angular2/bootstrap`). Application entry point (where your `main` method is) would import this library instead of the previously used `package:angular2/angular2.dart`. Non-main files should continue using `package:angular2/angular2.dart`. The new API no longer requires that user instantiate `ReflectionCapabilities` explicitly in their code.
### Example:

Before:

``` dart
import 'package:angular2/angular2.dart';
import 'package:angular2/src/reflection/reflection.dart' show reflector;
import 'package:angular2/src/reflection/reflection_capabilities.dart'
    show ReflectionCapabilities;
import 'package:myapp/myapp.dart';

main() {
  reflector.reflectionCapabilities = new ReflectionCapabilities();
  bootstrap(MyApp);
}
```

After:

``` dart
import 'package:angular2/bootstrap.dart';
import 'package:myapp/myapp.dart';

main() {
  bootstrap(MyApp);
}
```
## Breaking change

`package:angular2/angular2.dart` no longer provides the `bootstrap` method. This is because `bootstrap` pull in `dart:mirrors`, which we want to avoid in production code. Entry points are exempt because they are rewritten at build time from `bootstrap` to `bootstrapStatic`.
## Transformer

The transformer rewrites the above "after" example into the following mirror-free code:

``` dart
import 'package:angular2/bootstrap_static.dart';
import 'main.ng_deps.dart' as i0;
import 'package:myapp/myapp.dart';

main() {
  bootstrapStatic(MyApp, i0.initReflector);
}
```

### CODE REVIEW

**Comment by gdi2290** [comment]
is this the bootstrap refactor just to fix your issue or is it the full bootstrap refactor?
```es6
bootstrap(Type, componentBindings, appBindindings, platformBindings):Application {
  var platform = Platform.create(platformBindings);
  var app = platform.createApp(appBindindings);
  var componentRef = app.loadComponent(ComponentRef, componentBindings);
  return app;
}
```

**Comment by yjbanov** [comment]
@gdi2290, correct, this is to fix https://github.com/angular/angular/issues/2304 (remove reflection boilerplate in Dart). It does _not_ include https://github.com/angular/angular/issues/1594 (fluent bootstrap interface).

**Comment by kegluneq** [comment]
Looking good so far, thanks Yegor.

**Comment by gdi2290** [comment]
@yjbanov kool, does it make sense to promote bootstrap logic into a bootstrap folder `modules/angular2/src/bootstrap/`? The idea is that someone could manage their own dependencies if they didn't want to use di. This would also work in favor for future refactor since platform/appComponent/host folders could live there.

**Comment by yjbanov** [comment]
@gdi2290, it does make sense. I'll make a note of that and perhaps sneak in this change too.

File: modules/angular2/src/core/application.ts [hunk 1]
```diff
@@ -1,367 +1,7 @@
...
-    throw exception;
-  };
-
-  var reporter = isPresent(givenReporter) ? givenReporter : defaultErrorReporter;
-
-  var zone = new NgZone({enableLongStackTrace: assertionsEnabled()});
-  zone.overrideOnErrorHandler(reporter);
-  return zone;
-}
+import {commonBootstrap} from './application_common';
```
Comment (kegluneq) [inline modules/angular2/src/core/application.ts]: This is strictly a subset of `angular2.dart` -- can we remove it?

I know things might be directly referencing it (though they shouldn't!) -- maybe in a separate PR?

File: modules/angular2/src/transform/reflection_remover/rewriter.dart [hunk 2]
```diff
@@ -149,19 +77,131 @@ class _FindReflectionCapabilitiesVisitor extends Object
...
+    _currentIndex = node.end;
+  }
+
+  _rewriteBootstrapCallToStatic(MethodInvocation node) {
+    if (_rewriter._writeStaticInit) {
+      buf.write(_rewriter._code.substring(_currentIndex, node.offset));
+      _writeStaticReflectorInitOnce();
+      // rewrite `bootstrap(...)` to `bootstrapStatic(...)`
+      buf.write("bootstrapStatic");
+      buf.write(node.argumentList.toString());
```
Comment (kegluneq) [inline modules/angular2/src/transform/reflection_remover/rewriter.dart]: Consider the more idiomatic
```
buf.write('bootstrapStatic${node.argumentList}');
```

File: modules/angular2/src/core/application_static.dart [hunk 3]
```diff
@@ -0,0 +1,18 @@
+library angular2.application_static;
+
+import 'dart:async';
+import 'application_common.dart';
+
+export 'application_common.dart' show ApplicationRef;
+
+/// Starts an application from a root component. This implementation uses a
+/// pre-build function [initReflectorFn] as a source of reflection data instead
```
Comment (kegluneq) [inline modules/angular2/src/core/application_static.dart]: No longer accurate

File: modules/angular2/src/transform/reflection_remover/rewriter.dart [hunk 4]
```diff
@@ -149,19 +77,131 @@ class _FindReflectionCapabilitiesVisitor extends Object
...
         !reflectionCapabilityAssignments.contains(node.parent)) {
       logger.error('Unexpected format in creation of '
           '${REFLECTION_CAPABILITIES_NAME}');
     }
     return super.visitInstanceCreationExpression(node);
   }
+
+  @override
+  Object visitMethodInvocation(MethodInvocation node) {
+    if (node.methodName.toString() == 'bootstrap') {
```
Comment (kegluneq) [inline modules/angular2/src/transform/reflection_remover/rewriter.dart]:
```
if ('${node.methodName}' == BOOTSTRAP_NAME)
```

File: modules/angular2/src/transform/reflection_remover/rewriter.dart [hunk 5]
```diff
@@ -130,18 +45,31 @@ class Rewriter {
...
   Object visitImportDirective(ImportDirective node) {
-    if (_tester.isReflectionCapabilitiesImport(node)) {
-      reflectionCapabilityImports.add(node);
+    if (!_importAdded && _rewriter._writeStaticInit) {
+      // Add imports for ng_deps (once)
+      buf.write(_rewriter._code.substring(_currentIndex, node.offset));
+      buf.write(_rewriter._codegen.codegenImport());
+      _importAdded = true;
+      _currentIndex = node.offset;
+    }
```
Comment (yjbanov) [inline modules/angular2/src/transform/reflection_remover/rewriter.dart]: let's output ng_deps imports _after_ the first import

File: modules/angular2/src/transform/reflection_remover/rewriter.dart [hunk 6]
```diff
@@ -149,19 +77,131 @@ class _FindReflectionCapabilitiesVisitor extends Object
...
+      _rewriteBootstrapCallToStatic(node);
+    }
+    return super.visitMethodInvocation(node);
+  }
+
+  String outputRewrittenCode() {
+    if (_currentIndex < _rewriter._code.length) {
+      buf.write(_rewriter._code.substring(_currentIndex));
+    }
+    return buf.toString();
```
Comment (kegluneq) [inline modules/angular2/src/transform/reflection_remover/rewriter.dart]:
```
return '$buf';
```

**Comment by kegluneq** [comment]
lgtm pending js timeout issues -- thanks for taking on this big, gnarly refactor Yegor!

File: modules/angular2/src/transform/reflection_remover/rewriter.dart (hunk 4 above)
Comment (yjbanov) [inline modules/angular2/src/transform/reflection_remover/rewriter.dart]: done.

File: modules/angular2/src/transform/reflection_remover/rewriter.dart (hunk 6 above)
Comment (yjbanov) [inline modules/angular2/src/transform/reflection_remover/rewriter.dart]: done

File: modules/angular2/src/transform/reflection_remover/rewriter.dart (hunk 2 above)
Comment (yjbanov) [inline modules/angular2/src/transform/reflection_remover/rewriter.dart]: done

File: modules/angular2/src/core/application.ts (hunk 1 above)
Comment (yjbanov) [inline modules/angular2/src/core/application.ts]: A bigger cleanup of our exports is coming soon. Let's do it as part of that.

File: modules/angular2/src/core/application_static.dart (hunk 3 above)
Comment (yjbanov) [inline modules/angular2/src/core/application_static.dart]: fixed

File: modules/angular2/src/transform/reflection_remover/rewriter.dart (hunk 5 above)
Comment (yjbanov) [inline modules/angular2/src/transform/reflection_remover/rewriter.dart]: done

**Comment by yjbanov** [comment]
@kegluneq: re JS timeouts, @mhevery and I looked into it and there's no obvious fix; we flattened our exports, which resulted in some duplication, but solved the timeout issue; should be good for now.

**Comment by kegluneq** [comment]
SGTM, thanks!

**Pull request closed without merging**